board/safety/safety_ford.h

```diff
@@ -1,112 +1,219 @@
+addr_checks ford_rx_checks = {ford_addr_checks, FORD_ADDR_CHECK_LEN};
+
+
+static const addr_checks* ford_init(int16_t param) {
+  UNUSED(param);
 
+  controls_allowed = false;
+  relay_malfunction_reset();
+
+  return &ford_rx_checks;
+}
 
 static int ford_rx_hook(CANPacket_t *to_push) {
+  bool valid = addr_safety_check(to_push, &ford_rx_checks, NULL, NULL, NULL);
 
-  int addr = GET_ADDR(to_push);
-  int bus = GET_BUS(to_push);
-  bool unsafe_allow_gas = unsafe_mode & UNSAFE_DISABLE_DISENGAGE_ON_GAS;
+  if (valid && (GET_BUS(to_push) == FORD_MAIN)) {
+    int addr = GET_ADDR(to_push);
 
-  if (addr == 0x217) {
-    // wheel speeds are 14 bits every 16
-    vehicle_moving = false;
-    for (int i = 0; i < 8; i += 2) {
-      vehicle_moving |= GET_BYTE(to_push, i) | (GET_BYTE(to_push, (int)(i + 1)) & 0xFCU);
+    // Update steering angle
+    if (addr == MSG_STEERING_PINION_DATA) {
+      // Steering angle: (0.1 * val) - 3200 in deg
+      // Signal: StePinRelInit_An_Sns
+      int angle_meas_new = ((GET_BYTE(to_push, 0) << 8) | GET_BYTE(to_push, 1)) - 3200U;
+      update_sample(&angle_meas, angle_meas_new);
     }
-  }
 
-  // state machine to enter and exit controls
-  if (addr == 0x83) {
-    bool cancel = GET_BYTE(to_push, 1) & 0x1U;
-    bool set_or_resume = GET_BYTE(to_push, 3) & 0x30U;
-    if (cancel) {
-      controls_allowed = 0;
-    }
-    if (set_or_resume) {
-      controls_allowed = 1;
+    // Update in motion state from speed value
+    if (addr == MSG_ENG_VEHICLE_SP_THROTTLE2) {
+      // Speed in kph, convert to m/s
+      // Speed: (0.01 * val) in kph
+      // Signal: Veh_V_ActlEng
+      vehicle_speed = ((GET_BYTE(to_push, 6) << 8) | GET_BYTE(to_push, 7)) * 0.01 / 3.6;
+      vehicle_moving = vehicle_speed > 0.3;
     }
-  }
 
-  // exit controls on rising edge of brake press or on brake press when
-  // speed > 0
-  if (addr == 0x165) {
-    brake_pressed = GET_BYTE(to_push, 0) & 0x20U;
-    if (brake_pressed && (!brake_pressed_prev || vehicle_moving)) {
-      controls_allowed = 0;
+    // Update gas pedal
+    if (addr == MSG_ENG_VEHICLE_SP_THROTTLE) {
+      // Pedal position: (0.1 * val) in percent
+      // Signal: ApedPos_Pc_ActlArb
+      unsigned int raw_gas_can = ((GET_BYTE(to_push, 0) & 0x03U) << 8) | GET_BYTE(to_push, 1);
+      gas_pressed = raw_gas_can > 0U;
     }
-    brake_pressed_prev = brake_pressed;
-  }
 
-  // exit controls on rising edge of gas press
-  if (addr == 0x204) {
-    gas_pressed = ((GET_BYTE(to_push, 0) & 0x03U) | GET_BYTE(to_push, 1)) != 0U;
-    if (!unsafe_allow_gas && gas_pressed && !gas_pressed_prev) {
-      controls_allowed = 0;
+    // Update brake pedal and cruise state
+    if (addr == MSG_ENG_BRAKE_DATA) {
+      // Signal: BpedDrvAppl_D_Actl
+      brake_pressed = (GET_BYTE(to_push, 0) & 0x30U);
+
+      // Signal: CcStat_D_Actl
+      unsigned int cruise_state = (GET_BYTE(to_push, 1) & 0x07U);
+      bool cruise_engaged = (cruise_state != 0U) && (cruise_state != 3U);
```

Comment:
`cruise_state` will have eight potential values, some of which are undefined and some of which may be error or other abnormal states according to your DBC. Don't check for a couple known disengaged states and assume the rest are engaged, check for known and documented *engaged* states and assume the rest are disengaged.

Reply:
Fixed, ty